File: api/app/services/entity_extract.py
"""S4-B 핵심 엔티티 추출 (룰 기반 1차) — master plan §6 P1 (S4-B).

목적
----
chunks 의 metadata 에 핵심 엔티티 (날짜, 금액, 식별 번호, 비율) 자동 추출 →
search 시 단편 정확도 향상. 룰 기반 (정규식) 우선 — Flash-Lite LLM 보강은 별도 sprint.

설계 원칙
- **외부 의존성 0** — 정규식만 사용
- **graceful** — 매칭 실패 시 빈 list (raise X)
- **Korean-first** — 한국어 날짜/금액 패턴 우선 매칭
- **dedup** — 같은 string 중복 제거

추출 범위 (룰 기반)
- **dates**: 2024년 4월 30일 / 2024.04.30 / 2024-04-30 / 2024.4. / 24/4/30 등
- **amounts**: 1,000원 / 100만원 / $100 / 1억원 / 50% / 25.5% 등
- **identifiers**: 법령번호 / 보고서번호 / 표준번호 (ISSN, ISBN 등)
- **percentages**: 1.5%, 25%, 100%

LLM 보강 영역 (별도 sprint)
- 제품명 (예: "쏘나타 디 엣지", "Indigo Book")
- 기관명 (예: "한국은행", "한마음생활체육관")
- 인명 (예: "김뮤지", "아리아나")

사용
----
    from app.services.entity_extract import extract_entities

    text = "2024년 4월 30일부터 시행. 회비 50,000원 (50%) — 한국은행 ISSN 2288-7083"
    entities = extract_entities(text)
    # {"dates": ["2024년 4월 30일"], "amounts": ["50,000원"],
    #  "identifiers": ["2288-7083"], "percentages": ["50%"]}
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# 날짜 — 한국어 + ISO 형식 다양.
# `\b` 가 한국어 인접 시 매칭 불안정 (Korean char 가 \w 라 boundary X).
# → 한국어 패턴은 \b 제거 + greedy match 의존, ISO 패턴은 lookbehind/lookahead 사용.
_DATE_PATTERNS = [
    # 2024년 4월 30일 / 2024년 4월 (optional "일" group)
    re.compile(r"(\d{4})\s*년\s*(\d{1,2})\s*월(?:\s*(\d{1,2})\s*일)?"),
    # 2024.04.30 / 2024-04-30 / 2024/4/30 (ISO, ASCII boundary)
    re.compile(r"(?<!\d)(\d{4})[.\-/]\s*(\d{1,2})[.\-/]\s*(\d{1,2})(?!\d)"),
    # 2024.4 / 2024-04 (월까지만)
    re.compile(r"(?<!\d)(\d{4})[.\-]\s*(\d{1,2})(?!\d)"),
]

# 금액 — 1,000원 / 100만원 / 1억원 / $100 / ₩1,000
_AMOUNT_PATTERNS = [
    # 50,000원 / 1,234.56원 / 1,000,000원
    re.compile(r"\b\d{1,3}(?:,\d{3})+(?:\.\d+)?\s*원\b"),
    # 100만원 / 1억원 / 50조원
    re.compile(r"\b\d{1,4}(?:\.\d+)?\s*(?:만|억|조|천)\s*원\b"),
    # $100 / $1,000.50
    re.compile(r"\$\d{1,3}(?:,\d{3})*(?:\.\d+)?\b"),
    # ₩1,000
    re.compile(r"₩\d{1,3}(?:,\d{3})*(?:\.\d+)?\b"),
]

# 백분율 — 25%, 1.5%, 100%
_PERCENT_PATTERN = re.compile(r"\b\d{1,3}(?:\.\d+)?\s*%")

# 식별자 — ISSN/ISBN/법령번호 등
_IDENTIFIER_PATTERNS = [
    # ISSN 2288-7083
    re.compile(r"\bISSN\s+(\d{4}-\d{3}[\dXx])\b"),
    # ISBN 978-0-...
    re.compile(r"\bISBN\s+([\d\-Xx]{10,17})\b"),
    # 내규 제709호 / 법률 제12345호
    re.compile(r"제\s*\d+\s*호"),
]


@dataclass
class ExtractedEntities:
    dates: list[str]
    amounts: list[str]
    percentages: list[str]
    identifiers: list[str]

    def to_dict(self) -> dict[str, list[str]]:
        return {
            "dates": self.dates,
            "amounts": self.amounts,
            "percentages": self.percentages,
            "identifiers": self.identifiers,
        }

    def is_empty(self) -> bool:
        return not (
            self.dates or self.amounts or self.percentages or self.identifiers
        )


def _dedup_preserve_order(items: list[str]) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for it in items:
        s = it.strip()
        if s and s not in seen:
            seen.add(s)
            out.append(s)
    return out


def extract_dates(text: str) -> list[str]:
    out: list[str] = []
    if not text:
        return out
    for pattern in _DATE_PATTERNS:
        for m in pattern.finditer(text):
            out.append(m.group(0).strip())
    return _dedup_preserve_order(out)


def extract_amounts(text: str) -> list[str]:
    out: list[str] = []
    if not text:
        return out
    for pattern in _AMOUNT_PATTERNS:
        for m in pattern.finditer(text):
            out.append(m.group(0).strip())
    return _dedup_preserve_order(out)


def extract_percentages(text: str) -> list[str]:
    if not text:
        return []
    return _dedup_preserve_order(
        m.group(0).strip() for m in _PERCENT_PATTERN.finditer(text)
    )


def extract_identifiers(text: str) -> list[str]:
    out: list[str] = []
    if not text:
        return out
    for pattern in _IDENTIFIER_PATTERNS:
        for m in pattern.finditer(text):
            # group(1) 있으면 captured part, 아니면 전체 match
            value = m.group(1) if m.lastindex else m.group(0)
            out.append(value.strip())
    return _dedup_preserve_order(out)


def extract_entities(text: str) -> ExtractedEntities:
    """텍스트에서 룰 기반 엔티티 일괄 추출."""
    return ExtractedEntities(
        dates=extract_dates(text),
        amounts=extract_amounts(text),
        percentages=extract_percentages(text),
        identifiers=extract_identifiers(text),
    )

File: supabase/functions/_shared/ingest/entity_extract.ts
/**
 * 룰 기반 엔티티 추출 — `services/entity_extract.py` 의 정형 패턴 4종 포팅.
 *
 * `chunk` 가 청크마다 부른다(`chunks.metadata.entities`). LLM 보강 부분
 * (`persons`/`orgs`/`products`)은 chunk 경로에서 안 쓰므로 옮기지 않았다.
 *
 * ## `\b` 를 그대로 옮기면 한국어 금액이 통째로 안 잡힌다
 * Python `\b` 는 **유니코드 `\w`** 경계, JS `\b` 는 **ASCII `\w`** 경계다. 실측(2026-09-07):
 *
 * | 입력 | Python | JS `\b` |
 * |---|---|---|
 * | `50,000원` | 매칭 | **안 함** (`원` 이 JS 기준 non-word) |
 * | `100만원` | 매칭 | **안 함** |
 * | `$100달러` | 안 함 | **매칭** (`달` 이 JS 기준 non-word) |
 * | `약25%` | 안 함 | **매칭** |
 *
 * 그래서 `\b` 를 `PY_WORD_CLASS` lookaround 로 풀어 쓴다. **`v` 플래그가 필요하다**
 * (집합 뺄셈). `u` 와는 함께 못 쓴다.
 *
 * ## `\d` 도 다르다
 * Python `\d` 는 유니코드 Nd 전부 — `２５%`·`５０,０００원`(전각)이 잡힌다. `\p{Nd}` 로 옮겼다.
 */

import { PY_SP, pyStrip } from "../search/pystr.ts";
import { PY_WORD_CLASS } from "../pychar.ts";

/**
 * Python `\b` 그대로 — "한쪽은 word, 다른 쪽은 아님". 양쪽 lookbehind 실패가
 * 문자열 시작/끝을 자동으로 non-word 취급해 준다.
 *
 * `(?<!W)`(앞 경계) / `(?!W)`(뒤 경계) 로 줄여 쓰고 싶지만 **그건 패턴의 양끝 문자가
 * 반드시 word 일 때만 같다.** ISBN 패턴 `[\p{Nd}\-Xx]{10,17}` 은 `-` 로 끝날 수 있어
 * (`ISBN 1234567890-a` → Python 은 `1234567890-` 를 잡는다) 줄여 쓰면 갈린다.
 */
const B = `(?:(?<=${PY_WORD_CLASS})(?!${PY_WORD_CLASS})|(?<!${PY_WORD_CLASS})(?=${PY_WORD_CLASS}))`;
const B_BEFORE = B;
const B_AFTER = B;
/** Python `\d`(유니코드 Nd) · `\s`(Python 공백 집합). */
const D = "\\p{Nd}";
const S = `[${PY_SP}]`;

/** `v` 플래그로 만든다 — 집합 뺄셈이 들어 있다. */
const re = (src: string) => new RegExp(src, "gv");

/** 날짜 — `2024년 4월 30일` / `2024.04.30` / `2024-04` */
const DATE_PATTERNS = [
  re(`(${D}{4})${S}*년${S}*(${D}{1,2})${S}*월(?:${S}*(${D}{1,2})${S}*일)?`),
  re(`(?<!${D})(${D}{4})[.\\-\\/]${S}*(${D}{1,2})[.\\-\\/]${S}*(${D}{1,2})(?!${D})`),
  re(`(?<!${D})(${D}{4})[.\\-]${S}*(${D}{1,2})(?!${D})`),
];

/** 금액 — `50,000원` / `100만원` / `$100` / `₩1,000` */
const AMOUNT_PATTERNS = [
  re(`${B_BEFORE}${D}{1,3}(?:,${D}{3})+(?:\\.${D}+)?${S}*원${B_AFTER}`),
  re(`${B_BEFORE}${D}{1,4}(?:\\.${D}+)?${S}*(?:만|억|조|천)${S}*원${B_AFTER}`),
  re(`\\$${D}{1,3}(?:,${D}{3})*(?:\\.${D}+)?${B_AFTER}`),
  re(`₩${D}{1,3}(?:,${D}{3})*(?:\\.${D}+)?${B_AFTER}`),
];

/** 백분율 — `25%` / `1.5%` */
const PERCENT_PATTERN = re(`${B_BEFORE}${D}{1,3}(?:\\.${D}+)?${S}*%`);

/** 식별자 — ISSN / ISBN / `제709호` */
const IDENTIFIER_PATTERNS = [
  re(`${B_BEFORE}ISSN${S}+(${D}{4}-${D}{3}[${D}Xx])${B_AFTER}`),
  re(`${B_BEFORE}ISBN${S}+([${D}\\-Xx]{10,17})${B_AFTER}`),
  re(`제${S}*${D}+${S}*호`),
];

export interface ExtractedEntities {
  dates: string[];
  amounts: string[];
  percentages: string[];
  identifiers: string[];
}

/**
 * strip 후 빈 것 제외, **순서 보존** 중복 제거.
 *
 * `pyStrip` 을 `trim()` 으로 되돌려도 음성 대조가 0 건이었다. 케이스 부족이 아니라
 * **구조적 no-op** 이다 — 패턴 8 개가 전부 숫자·`$`·`₩`·`제` 로 시작해 숫자·`원`·`%`·
 * `호`·`[Nd Xx]` 로 끝나므로 매칭 결과가 공백으로 시작하거나 끝날 수 없다(무작위
 * 200,000 입력 × 30,587 매칭에서 `strip()` 이 값을 바꾼 경우 0 건). 원본 계약이라
 * 남겨 두되, 여기서 갈릴 일은 없다.
 */
function dedupPreserveOrder(items: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const it of items) {
    const s = pyStrip(it);
    if (s && !seen.has(s)) {
      seen.add(s);
      out.push(s);
    }
  }
  return out;
}

function collect(text: string, patterns: RegExp[], useGroup1 = false): string[] {
  if (!text) return [];
  const out: string[] = [];
  for (const p of patterns) {
    p.lastIndex = 0; // `g` 정규식은 상태를 갖는다 — 재사용 전에 반드시 되돌린다.
    for (const m of text.matchAll(p)) {
      // 원본 `m.group(1) if m.lastindex else m.group(0)` —
      // 그룹이 **매칭됐을 때만** 1 번을 쓴다.
      out.push(useGroup1 && m[1] !== undefined ? m[1] : m[0]);
    }
  }
  return dedupPreserveOrder(out);
}

export function extractDates(text: string): string[] {
  return collect(text, DATE_PATTERNS);
}
export function extractAmounts(text: string): string[] {
  return collect(text, AMOUNT_PATTERNS);
}
export function extractPercentages(text: string): string[] {
  return collect(text, [PERCENT_PATTERN]);
}
export function extractIdentifiers(text: string): string[] {
  return collect(text, IDENTIFIER_PATTERNS, true);
}

export function extractEntities(text: string): ExtractedEntities {
  return {
    dates: extractDates(text),
    amounts: extractAmounts(text),
    percentages: extractPercentages(text),
    identifiers: extractIdentifiers(text),
  };
}

/** 네 항목이 전부 비면 참. 호출부는 이때 `metadata.entities` 를 아예 안 넣는다. */
export function entitiesEmpty(e: ExtractedEntities): boolean {
  return !(e.dates.length || e.amounts.length || e.percentages.length ||
    e.identifiers.length);
}
